Add method KeypointBasedMotionEstimator::estimate(InputArray, InputArray) to support both CPU & OpenCL processing

zongwave opened the pull request:
the original estimate function's input parameters is defined as "Mat", this prevent users to call into algorithm opencl path

<!-- Please use this line to close one or multiple issues when this pullrequest gets merged
You can add another line right under the first one:
resolves #1234
resolves #1235
-->

### This pullrequest changes

<!-- Please describe what your pullrequest is changing -->

1. it's impossible to call into openCL implementation of algorithm "feature detect and optical flow" through the original function KeypointBasedMotionEstimator::estimate(const Mat &frame0, const Mat &frame1, bool *ok = 0) 

the parameter "Mat" passed into feature detect and optical flow algorithm will choose Non-opencl path. add a new function with input parameters "UMat", so that if user can choose run the algorithum with opencl.

2. ocl_calcOpticalFlowPyrLK function only accept 1 channel gray image, so have to convert input image format into one channel grey image if it is necessary.

vpisarev commented:
@zongwave, thank you for the contribution!

The overhead of copying Mat to UMat in this algorithm should be much lower than the processing time, so probably it makes sense to put OpenCL branch into the regular estimate() method and add some method that would control OpenCL use by the algorithm.

zongwave commented:
@vpisarev I updated the patch according to your comments, could you please help review the patch?

zongwave commented:
@vpisarev  our project under developing is depending on this change. could you mind help review the patch? since I'v updated the change according to your comments.
Thank you very much.

vpisarev commented:
sorry for delay; let's put it in :+1:

alalek commented:
I believe, we should not execute OpenCL code without explicit usage of `cv::UMat` somewhere in the user application.

`Mat& frame` must be replaced to `InputArray` to support UMat arguments (but there is ABI issues).

BTW, class `KeypointBasedMotionEstimator` has a lot of virtual methods, but I can't find any derived class in OpenCV code.

zongwave commented:
@alalek @vpisarev thanks for your comments.
I've modified this patch twice, the first version I added one overload function in class KeypointBasedMotionEstimator
virtual Mat estimate(InputArray points0, InputArray points1, bool *ok = 0);
But most contents of this new function are redundant with the original function estimate()

What's your opinion on how to modify this patch in a decent way? I'm so appropriate your advice.
Thank you very much!

zongwave commented:
@alalek @vpisarev I updated this patch according to both your suggestions, do you think if it is OK?
